### PR TITLE
Add Scanner for YR_SCANNER

### PR DESCRIPTION
**Pull request by mble**
Hello!

Inspired by @plusvic's comment here: https://github.com/hillu/go-yara/pull/21#issuecomment-362015462 and my own requirements (concurrent access to a `Rules` object with different external variables set), this PR implements an interface to `YR_SCANNER`, behaving much like the existing `YR_RULES` interface.

In order to provide an easier drop in wrapper for `Rules`, rather than having to explicitly call `s.AddCallback()` and friends, this is done in the initialiser in this version. Should the goal of the project to be closely mapping to the C API, happy to pull these out into their own methods on `Scanner`.

An additional stress test has also been added to make sure the concurrent access problem doesn't arise, as it does with `Rules`. A test to check `Scanner` independence was also added, lifted shamelessly from https://github.com/VirusTotal/yara/blob/master/tests/test-api.c#L354.

### CODE REVIEW

File: scanner.go [hunk 1]
```diff
@@ -36,6 +36,7 @@ import (
 // Scanner contains a YARA scanner
 type Scanner struct {
 	*scanner
+	rules *Rules
```
Comment (hillu) [inline scanner.go]: There should be a comment that explains that the rules object might be garbage-collected and cause the underlying YR_RULES struct to be destroyed if this field were missing.

**Comment by hillu** [comment]
@mble Thank you very much! Since I originally started that discussion about YARA's thread-safety issues that led to the creation of the Scanner interface, I really should have implemented this earlier in go-yara…

I would like to keep things close to the C API (simple option setting methods), will make a suggestion based on your branch.

File: scanner.go (hunk 1 above)
Comment (mble) [reply scanner.go]: Ack, will do!

**Comment by mble** [comment]
@hillu I've added a comment, and looked at setting the option setting methods – the main sticking point was a `SetCallback` method on `Scanner`: requiring the caller to implement their own handling of the callback, including deferring deleting the callback id from the callback pool, led to same awkward ergonomics.

I can appreciate wanting to expose these mappings, but for the sake of ergonomics, I think I'll keep these abstracted, like they are in the corresponding `Rules` API (at least in my fork).

**Comment by hillu** [comment]
@mble Would you like to take a look at the changes I have made in https://github.com/hillu/go-yara/tree/add-yr-scanner-2 ?

I think that SetXXX methods that can be chained are a sensible alternative to having to pass every option to the ScanXXX methods. (Sure, this is not exactly consistent with the (*Rules).ScanXXX methods.) Storing the currently set callback-id with the Scanner instance means that the API's user does not need to worry about that implementation detail at all and it can be cleaned up automatically when the Scanner is garbage-collected.

**Comment by mble** [comment]
@hillu Ah, I like the chained methods. I think the `ScanXXX2` methods are a nice convenience. My only critique would be to remove the commented out `r.DefineVariable` calls – otherwise looks great!

**Comment by hillu** [comment]
@mble Thanks!

I still don't really like the names of the convenience methods – do you have a suggestion for better names?

**Comment by mble** [comment]
@hillu Maybe something like `ScanMemWithDefaults`? I'm finding it tricky to capture semantically the intent behind them, and make sure they are explicit enough without being too wordy.

**Comment by hillu** [comment]
@mble I have come up with a simpler solution that puts everything in one ScanXxxx variant. I hope you approve. :-)
